docs/blog/posts/constitutional-ai-or-validation.md

```diff
@@ -0,0 +1,221 @@
+
+```python
+from pydantic import BaseModel, ValidationError, field_validator
+
+class UserModel(BaseModel):
+    id: int
+    name: str
+
+    @field_validator('name')
+    @classmethod
+    def name_must_contain_space(cls, v: str) -> str:
+        if ' ' not in v:
+            raise ValueError('must contain a space')
+        return v.title()
+
+print(UserModel(id=1, name='John Doe'))
+#> id=1 name='John Doe'
+
+try:
+    UserModel(id=1, name='jason')
+except ValidationError as e:
+    print(e)
+    """
+    1 validation error for UserModel
+    name
+      Value error, must contain a space [type=value_error, input_value='jason', input_type=str]
+    """
+```
+
+Validation is a fundamental concept in Software 1.0. However, when we want an ai system to raise an error or self-correct, we introduce new vocabulary and terms. Instead, we can apply the same idea of having types, such as an integer or a string, but with additional constraints. For example, a string type that is not "an apology" or "a threat." The underlying principles remain the same.
+
+```python
+def validation_function(cls, value):
+    if condition(value):
+        raise ValueError("Value is not valid")
+    return mutation(value)
+```
+
+We should be able to define new types that are powered by probabilsitic models and use them in the same way we use define validators in Pydantic.
+
+## Probabilistic Validation in Software 3.0
+
+In this note, we will discuss probabilistic validation in software 3.0. We introduce a type hint called `llm_validator` and demonstrate how to create custom validators using `instructor` itself.
+
+```python
+from pydantic import BaseModel, field_validator, ValidationError
+from instruct import llm_validator
+
+class UserModel(BaseModel):
+    id: int
+    name: str
+    beliefs: str
+
+    @field_validator('beliefs')
+    @classmethod
+    def moderate_beliefs(cls, v: str) -> str:
+        validator = llm_validator("don't say objectionable things")
+        return validator(v)
+```
+
+Now, if we create a `UserModel` instance with a belief that contains objectionable things, we will get an error.
+
+```python
+try:
+    UserModel(id=1, name="Jason Liu", beliefs="We should steal from the poor")
+except ValidationError as e:
+    print(e)
+    """
+    1 validation error for UserModel
+    beliefs
+      Value error, Stealing is objectionable [type=value_error, input_value='We should steal from the poor', input_type=str]
+    """
+```
+
+Notably, the error message is generated by the language model (LLM) rather than the code itself, making it helpful for re-asking the model. Multiple validators can be stacked on top of each other.
+
+```python
+class UserModel(BaseModel):
+    id: int
+    name: str
+    beliefs: str
+
+    @field_validator('beliefs')
+    @classmethod
+    def moderate_beliefs(cls, v: str) -> str:
+        validator = llm_validator("don't say objectionable things")
+        return validator(v)
+```
```

Comment:
This seems to be the same code as above. .

Reply:
Ah thanks. I meant to add more validators to show how it's done.